### PR TITLE
Introduce rustc_interface and move some methods there

### PR DESCRIPTION
**Pull request by Zoxc**
Split out from https://github.com/rust-lang/rust/pull/56732

r? @oli-obk

### CODE REVIEW

File: src/librustc_interface/util.rs [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+        .crate_types
+        .borrow()
+        .contains(&config::CrateType::Executable);
+    hasher.write(if is_exe { b"exe" } else { b"lib" });
+
+    CrateDisambiguator::from(hasher.finish())
+}
+
+pub fn collect_crate_types(session: &Session, attrs: &[ast::Attribute]) -> Vec<config::CrateType> {
+    // Unconditionally collect crate types from attributes to make them used
```
Comment (phansch) [inline src/librustc_interface/util.rs]: Should this be a doc comment?

Comment (Zoxc) [reply src/librustc_interface/util.rs]: I don't think so

**Review by oli-obk — COMMENTED** [review]
r=me with nits fixed

File: src/librustc_driver/driver.rs [hunk 2]
```diff
@@ -1222,100 +1221,9 @@ where
...
-            });
-
-            time(sess, "privacy checking", || {
-                rustc_privacy::check_crate(tcx)
-            });
-
-            time(sess, "intrinsic checking", || {
-                middle::intrinsicck::check_crate(tcx)
-            });
+            tcx.analysis(LOCAL_CRATE).ok();
```
Comment (bjorn3) [inline src/librustc_driver/driver.rs]: This should return the error, shouldnt't it.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/librustc/ty/context.rs [hunk 3]
```diff
@@ -1282,8 +1282,6 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 
         let gcx = arenas.global_ctxt.as_ref().unwrap();
 
-        sync::assert_send_val(&gcx);
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc/ty/context.rs]: It was there after we modified `GlobalCtxt` to be `Sync`, but didn't yet actually share it across threads.

File: src/librustc_driver/driver.rs (hunk 2 above)
Comment (Zoxc) [reply src/librustc_driver/driver.rs]: We pass the compile status to the closure and expect it to handle it. I'll change `phase_3_run_analysis_passes` to return just `R`.

**Comment by Centril** [comment]
@Zoxc Can you please follow this PR up with one that converts the new crate to Rust 2018?

**Comment by Zoxc** [comment]
@Centril Maybe after https://github.com/rust-lang/rust/pull/56732 lands

**Comment by Centril** [comment]
Too big to rollup, @bors p=1

**Comment by Centril** [comment]
@bors retry

yielding to rollup

**Pull request merged**